RFC: approach to supporting driver-specific API

pabigot opened the pull request:
This PR serves as an example supporting a proposed standard approach to driver-specific APIs (#11993).  Please review the discussion in that issue.  This PR is designed to present an example for discussion and is *not* intended to be merged, ever.  PR #17631 provides details on a specific case where this is needed.

**NOTE** The discussion here starts from the position that some devices provide extended functionality for which access is desirable but a generic API is simply not feasible.  To avoid splitting the discussion please pose concerns about this assumption and other generic concerns in #11993 where they can  reference points already made there.

The proposed technical approach involves the device using [a driver-specific header](https://github.com/pabigot/zephyr/blob/dpr/11993/include/drivers/rtc/ds3231.h), and an API table that [contains the API from the base system (e.g. counter, sensor), but adds functions for its own functions](https://github.com/pabigot/zephyr/blob/fadebb02f5bbdd24684078c39b9c57502f712d43/drivers/counter/counter_ds3231.c#L186).  This approach is in contrast to use of some sort of generic `ioctl(2)` API, or to a per-subsystem `cmd` function.

There is also a minimal [example](https://github.com/pabigot/zephyr/blob/fadebb02f5bbdd24684078c39b9c57502f712d43/samples/drivers/rtc_ds3231/src/main.c#L20) that shows how the device can be accessed through its generic and specific APIs.

It's understood that such API will only work on applications that target platforms supporting the corresponding device.  Should another device with similar features turn up it is expected that developers will re-evaluate whether a generic API is suitable.

Currently Zephyr does limited/no validation that a passed `struct device*` is correctly paired with the API function it's invoking.  The approach here does not change that.  Material from #12056 may support such validation, though the mechanism for using it is unclear and the solution would likely only work for applications that enable userspace.

The material presented here is extracted from work-in-progress to support the Maxim DS3231 temperature-compensated Real Time Clock, allowing a standalone Zephyr device to be able to map observed events and behaviors between system time and civil time.  The best existing driver model for this device is counter, but as an external I2C device it may not satisfy many expectations of the common API, and full use requires operations that cannot be expressed in that API (e.g., setting an alarm that first at 11:03 am every Thursday, or on the 24th of the month at midnight).

All material related to the operations has been removed to reduce clutter, but the API documentation (which is also WIP) has been retained in hopes they support the position that the operations they support are not suitable for generalization at this time.

pabigot committed:
DNM WIP drivers: counter: add Maxim DS3231 support

This is a stub implementation to be used as the basis for discussion
of solutions to issue #11993.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
Possibly relevant: the in-tree Grove LCD display driver does not support the Zephyr Display API, but instead implements its own functions bypassing the system call approach.

nordic-krch commented on include/drivers/rtc/ds3231.h:
```diff
@@ -0,0 +1,398 @@
+/*
```
i wonder about the location of this file. implementation is in `drivers/counter`. What about putting it in `include/drivers/counter`?

pabigot replied:
That's somewhat specific to the example, but not entirely, so I'll outline the general issue.

Fundamentally the DS3231 is a real-time-clock device, not a counter.  It's a counter only because that's the closest existing generic driver class.  The source is in `counter` because it implements that API, but anybody who wants to use it will be operating near exclusively with the device-specific functions.

IMO using `#include <drivers/counter/ds3231.h>` will mislead people into thinking it really is a counter.  But DS3231 can support only extremely limited use of the counter API, and some things that are expected (like being able to read from within an ISR) are totally unsupported.  A strong case can be made that it shouldn't be a counter at all.

But that opens up the question of where to put devices that have no relation to any generic driver class.  Or, to whether in this case we should bring back an RTC driver API that provides the functionality required for using an RTC separately from counter.

Discussion about that probably belongs in the PR for the DS3231 module though.  In the other driver for this sort of thing--extensions to the sensor API--the devices fully support the base driver class, and simply require a few extra functions to expose their full functionality.  Those would go into `include/drivers/sensors/abc321z.h`.

nordic-krch replied:
i agree, that this case is specific. In case of driver which fits to specific api and extends it, there will be no such problem.

MaureenHelm replied:
Implementing the generic counter interface will make people think it's a counter. If you're worried about people thinking it's a counter, then why implement the generic counter interface at all?

Similarly, if you think applications will use the device-specific functions exclusively, then isn't implementing the generic counter interface a waste of code space? I think we either need to revisit the RTC driver API or you should have a driver-specific API.

That said, I think this is nice way to extend a generic driver API if that functionality is needed. I'm just not sure that DS3231 specifically is a good use case for it.

nordic-krch reviewed (commented):
it seems like a good approach. I wonder how to make it a standard?

bfrog commented on include/drivers/rtc/ds3231.h:
```diff
@@ -0,0 +1,398 @@
+					struct k_poll_signal *signal);
+
+struct rtc_ds3231_driver_api {
+	struct counter_driver_api counter_api;
```
This is a neat way of extending an existing API, and in some instances makes a lot of sense

teburd reviewed (commented):
This is an interesting device. I wonder if it even makes sense to have it implement the counter api like you yourself mention. If not, its confusing where it should go.

There are unique devices out there, it would be nice if zephyr supported them.

galak commented:
DEV-REVIEW: Marking this as 'TSC' with regards to the approach for how to have driver specific api's.  Intent is a next TSC meeting to vote on this.

pabigot committed:
doc/reference/drivers: document device-specific API extensions

Document the technical solution to providing a driver instance that
extends the functionality of a subsystem API.

Relates-to: #11993
Signed-off-by: Peter A. Bigot <pab@pabigot.com>

erwango commented:
I found the approach is good, and feature indeed required

pabigot commented:
Per [TSC Vote](https://docs.google.com/document/d/12p8Q4USdExsP3a8lIdhyqG-4vxb44kisccZ9TvlxWBA/edit#heading=h.ngp7j4uz7xgf) 2019-08-07 this approach is approved.

tomhepworth commented:
Hi, the links seem broken but I would love to see the examples you mention. Are they available anywhere?
